src/mallocTracer.cpp

```diff
@@ -0,0 +1,248 @@
+        return free_hook(addr);
+    }
+    return _orig_free(addr);
+}
+#endif
+
+u64 MallocTracer::_interval;
+volatile u64 MallocTracer::_allocated_bytes;
+
+Mutex MallocTracer::_patch_lock;
+int MallocTracer::_patched_libs = 0;
+bool MallocTracer::_initialized = false;
+
+// Only to be used from getOrigAddresses
+void* safeCalloc(size_t num, size_t size) {
+    void* ret = OS::safeAlloc(num * size);
+    if (likely(ret)) {
+        memset(ret, 0, num * size);
+    }
+    return ret;
+}
+
+__attribute__((constructor)) static void getOrigAddresses() {
+    // Set malloc and calloc which may be called from libc during getOrigAddresses.
+    _orig_malloc = OS::safeAlloc;
+    _orig_calloc = safeCalloc;
+
+    // Store these addresses, regardless of MallocTracer being enabled or not.
+    _orig_malloc = ADDRESS_OF(malloc);
+    _orig_calloc = ADDRESS_OF(calloc);
+    _orig_realloc = ADDRESS_OF(realloc);
+    _orig_free = ADDRESS_OF(free);
+}
+
+bool MallocTracer::initialize() {
+    if (!__sync_bool_compare_and_swap(&_initialized, false, true)) {
+        return false;
+    }
+
+    CodeCache* lib = Profiler::instance()->findLibraryByAddress((void*)MallocTracer::initialize);
+    assert(lib);
+
+    lib->mark(
+        [](const char* s) -> bool {
+            return strncmp(s, "_ZL11malloc_hook", 16) == 0
+                || strncmp(s, "_ZL11calloc_hook", 16) == 0
+                || strncmp(s, "_ZL12realloc_hook", 17) == 0
+                || strncmp(s, "_ZL9free_hook", 13) == 0;
+        },
+        MARK_ASYNC_PROFILER);
+
+    return installHooks();
+}
+
+bool MallocTracer::patchLibs(bool install) {
+    if (!initialized()) {
+        return false;
+    }
+
+    MutexLocker ml(_patch_lock);
+    if (!install) {
+        assert(_orig_malloc);
+        assert(_orig_calloc);
+        assert(_orig_realloc);
+        assert(_orig_free);
+
+        _patched_libs = 0;
+    }
+
+    CodeCacheArray* native_libs = Profiler::instance()->nativeLibs();
+    int native_lib_count = native_libs->count();
+
+    while (_patched_libs < native_lib_count) {
+        CodeCache* cc = (*native_libs)[_patched_libs++];
+
+        cc->patchImport(im_malloc, (void*)(install ? malloc_hook : _orig_malloc));
+        cc->patchImport(im_calloc, (void*)(install ? calloc_hook : _orig_calloc));
+        cc->patchImport(im_realloc, (void*)(install ? realloc_hook : _orig_realloc));
+        cc->patchImport(im_free, (void*)(install ? free_hook : _orig_free));
+    }
+
+    if (!install) {
+        _patched_libs = 0;
+    }
+
+    return true;
+}
+
+void MallocTracer::recordMalloc(void* address, size_t size) {
+    if (updateCounter(_allocated_bytes, size, _interval)) {
+        MallocEvent event;
+        event._start_time = OS::nanotime();
+        event._address = (uintptr_t)address;
+        event._size = size;
+
+        Profiler::instance()->recordSample(NULL, size, MALLOC_SAMPLE, &event);
+    }
+}
+
+void MallocTracer::recordFree(void* address) {
```

Comment:
Am I correct in saying we do not have a sampling decision here (as we do not want to miss matching frees) ?
[Here](https://github.com/DataDog/ddprof/blob/bccb05d89ec9e8473cd7f4f26d4e4948a70839c8/src/lib/allocation_tracker.cc#L275) we proposed to record only free events for tracked addresses. However that meant having a lock free data structure containing tracked addresses.
The ideal solution might be to have more control over the allocator being used, to embed such a logic within the allocated object.

Reply:
Yes, we do not have a sampling decision for `free` and wanted to avoid having to keep a list of allocation addresses.